### PR TITLE
New feature: interactive shell

### PR DESCRIPTION
**Pull request by KeenS**
*Issue #, if available:*

*Description of changes:*
This patch adds a new shell flag namely `--shell` . When `--shell` is passed without subcommands, dynein starts an interactive shell session. In the shell session, you can input any dynein subcommands and built-in shell command (currently,  only `exit` is defined to exit the shell).

```
$ dy --shell
> list
  hoge
  fuga
  piyo
> use fuga
Now you're using the table 'fuga' (ap-northeast-1);
> describe
---
name: fuga
region: ap-northeast-1
status: ACTIVE
....
> exit
```

As a bonus, you can  'script' the subcommands.

```console
$ cat script.txt
use fuga
describe
$ dy --shell < script.txt
...
```

How do you like this?

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -30,6 +30,7 @@ name = "dy"
 path = "src/main.rs"
 
 [dependencies]
+atty            = "0.2"
```
Comment (KeenS) [inline Cargo.toml:33]: This https://crates.io/crates/atty.
MIT licensed.

File: src/shell.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        debug!("Line read: {:?}", line);
+
+        match line {
+            // build-in shell command(s)
+            "exit" => Ok(ShellInput::Builtin(BuiltinCommands::Exit)),
+            // dy commands
+            line => {
+                // TODO: better handling of whitespaces
+                let args = line.split(' ');
```
Comment (KeenS) [inline src/shell.rs:57]: leaving it as naive. What syntax do you like (such as `"hoge fuga\npiyo"` or `'hoge fuga'\n'piyo'`)?

**Comment by thash** [comment]
Awesome feature! Thanks a lot for submitting the PR.
To make this feature nicer, let me discuss details of this interactive shell mode.

The first thing I noticed is that `use` command in the shell mode:

```
$ git checkout KeenS/shell-mode
$ cat .git/HEAD
[a7c6acca7f9faf3f17af728da3b08294602f5461](https://github.com/KeenS/dynein/commit/a7c6acca7f9faf3f17af728da3b08294602f5461)
$ cargo build --release
$ rm -rf ~/.dynein/
$ ./target/release/dy --shell
> config dump
---
tables: ~
---
using_region: ~
using_table: ~
> ls
DynamoDB tables in region: eu-central-1
  fuga
  hoge
> use fuga
Now you're using the table 'fuga' (eu-central-1).
> config dump
---
tables:
  eu-central-1/fuga:
    region: eu-central-1
    name: fuga
    ...
---
using_region: eu-central-1
using_table: fuga
> ls
DynamoDB tables in region: eu-central-1
  fuga
  hoge
> describe
[2020-11-28T13:41:02Z ERROR dy::app]
    To execute the command you must specify target table in one of following ways:
    ...
```

After `use` "fuga" table I expected that `ls` shows a line `* fuga` and dynein switches target table to "fuga". However, although config files are updated successfully, it seems that dynein doesn't switch the target table in shell mode.

Once you exit from the interactive shell mode and enter into the mode again, I could use "fuga" table appropriately like below:

```
// (continued from the previous code block) //
$ ./target/release/dy --shell
> ls
DynamoDB tables in region: eu-central-1
* fuga
  hoge
> desc
---
name: fuga
region: eu-central-1
status: ACTIVE
...
```

I guess the cause of the issue is that dynein loads config files only when the `dy` command is executed, whereas the shell mode continues to run in [the loop](https://github.com/awslabs/dynein/pull/13/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR260).

Looking at the description in the PR, I think this behavior is not what you intended to, but please let me know if I misunderstand expected behavior of the feature.

File: Cargo.toml (hunk 1 above)
Comment (thash) [reply Cargo.toml:33]: Sure, LGTM :)

File: src/shell.rs (hunk 2 above)
Comment (thash) [reply src/shell.rs:79]: I understand that current implementation requires a single whitespace as an argument separator. With regard to the multi-line syntax such as `"hoge fuga\npiyo"`, as the current dynein command is supposed to be passed on a single line, we do not have a specific policy on the syntax of passing arguments on multiple lines.

Possible syntaxes I can imagine would be (let me know if you have others promising options):
1. Accepts arguments with multiple lines and a command ends with semicolon(`;`) or other special character (like MySQL interactive shell).
2. Only when you put `\` at the end of line, the shell mode accepts next line (like `bash/zsh`).

I prefer No.2, but it's also ok to leave it as it is for now and work on it in future enhancements.

**Comment by KeenS** [comment]
Hmm...
Perhaps I was confused while doing try-and-error . 
I'll fix it as the way you mentioned and write test against it.

File: src/shell.rs (hunk 2 above)
Comment (KeenS) [reply src/shell.rs:79]: I mentioned how to include whitespaces in *tokens* but you are saying how to write multiple-*line*  inputs.
Both are important, but should be discussed separately.

On how to write multiple lines, I'm for 2 too.

Concerning how to include whitespaces in tokens, consider this input:
`query --sort-key '= 12' pk`
This is parsed as `query`, `--sort-key`, `'=`, `12'`, `pk` in current implementation.
This is not intended behaviour. We need some quotation syntaxes. However, there are options which syntax to adopt.
`'= 12'` ? Or `"= 12"`? When it comes to include quotation marks, how to escape them?
The first example I showed wasn't sufficient to describe this option. I apologize for that.

Then, which syntax do you prefer? Or leave it as-is and discus the syntax in some new issues?
IMO multi-line syntax can be postponed but minimal implementation of quoting is mandatory.

File: src/main.rs [hunk 3]
```diff
@@ -32,38 +32,21 @@ mod bootstrap;
...
-        config: Some(app::load_or_touch_config_file(true)?),
-        cache: Some(app::load_or_touch_cache_file(true)?),
-        overwritten_region: app::region_from_str(c.region),
-        overwritten_table_name: c.table,
-        output: None,
-    };
-    debug!("Initial command context: {:?}", &context);
-
-    match c.child {
+async fn dispatch(context: &mut app::Context, subcommand: cmd::Sub) -> Result<(), Box<dyn Error>> {
```
Comment (KeenS) [inline src/main.rs:42]: Changed `context` from `app::Context` to `&mut app::Context` and put `clone()` around.
Eliminating `clone()`s need further refactoring thus I won't do in this P-R.
I guess it is related to #6 and should be done in that fix.

File: src/shell.rs (hunk 2 above)
Comment (thash) [reply src/shell.rs:79]: I appreciate you describing it in detail, sorry for my confusion. Now I understand your query and I agree with you that we need to make a minimum decision about the issue on whitespaces in tokens here.

I prefer single-quotes `'= 12'` than double-quotes `"= 12"`.
Then, if we choose single-quotes, we need to decide how to escape single-quotes in them. One naive idea is just escaping it with backquotes (e.g. `> get 'I\'m a hoge'`)... Do you think this method will work?

File: src/main.rs (hunk 3 above)
Comment (thash) [reply src/main.rs:42]: Yes, in #6 we'll deal with the "too much `clone()`s" issue. So it's ok to add `clone()`s to make the feature work in this PR.

File: src/shell.rs (hunk 2 above)
Comment (KeenS) [reply src/shell.rs:79]: Yes, and I'll implement it.
I have a question. What syntax do you like to escape `\` itself? I think `\\` would be natural. If it looks good to you, I'll implement it too.

Comment (thash) [reply src/shell.rs:79]: Yes, using `\\` to escape `\` itself sounds natural for me as well :)

**Comment by KeenS** [comment]
Implemented quoting.

```
$ cargo run -- --shell
   Compiling dynein v0.1.2 (/home/shun/Rust/dynein)
    Finished dev [unoptimized + debuginfo] target(s) in 6.91s
     Running `target/debug/dy --shell`
> hoge fuga
Invalid argument: error: Found argument 'hoge' which wasn't expected, or isn't valid in this context

USAGE:
    dynein <SUBCOMMAND>

For more information try --help

> input isn't 'correct
Error while parsing input: quote isn't closed
> escabe isn't 'correct\
Error while parsing input: escape('\') is incomplete
>
```

@thash I guess all the comments are solved. Please re-review.

**Comment by thash** [comment]
I've reviewed changes and it looks good to me. Also I could confirm that the feature works fine - I would merge it.

**Pull request merged**